components/keys/src/types.rs

```diff
@@ -211,14 +212,126 @@ impl Display for Key {
     }
 }
 
+#[derive(Clone, Copy, Debug, Default, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct TimeStamp(u64);
+
+const TSO_PHYSICAL_SHIFT_BITS: u64 = 18;
+
+impl TimeStamp {
+    /// Create a time stamp from physical and logical components.
+    pub fn compose(physical: u64, logical: u64) -> TimeStamp {
+        TimeStamp((physical << TSO_PHYSICAL_SHIFT_BITS) + logical)
+    }
+
+    pub const fn min() -> TimeStamp {
+        TimeStamp(0)
+    }
+
+    pub const fn max() -> TimeStamp {
+        TimeStamp(std::u64::MAX)
+    }
+
+    pub const fn new(ts: u64) -> TimeStamp {
+        TimeStamp(ts)
+    }
+
+    /// Extracts physical part of a timestamp, in milliseconds.
+    pub fn extract_physical(self) -> u64 {
+        self.0 >> TSO_PHYSICAL_SHIFT_BITS
+    }
+
+    pub fn incr(self) -> TimeStamp {
+        TimeStamp(self.0 + 1)
+    }
+
+    pub fn decr(self) -> TimeStamp {
+        TimeStamp(self.0 - 1)
+    }
+
+    pub fn is_zero(self) -> bool {
+        self.0 == 0
+    }
+
+    pub fn into_inner(self) -> u64 {
+        self.0
+    }
+}
+
+impl From<u64> for TimeStamp {
+    fn from(ts: u64) -> TimeStamp {
+        TimeStamp(ts)
+    }
+}
+
+impl From<&u64> for TimeStamp {
+    fn from(ts: &u64) -> TimeStamp {
+        TimeStamp(*ts)
+    }
+}
+
+impl fmt::Display for TimeStamp {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        Display::fmt(&self.0, f)
+    }
+}
+
+impl slog::Value for TimeStamp {
+    fn serialize(
+        &self,
+        record: &slog::Record,
+        key: slog::Key,
+        serializer: &mut dyn slog::Serializer,
+    ) -> slog::Result {
+        slog::Value::serialize(&self.0, record, key, serializer)
+    }
+}
+
+/// A time in seconds since the start of the Unix epoch.
+#[derive(Clone, Copy, Debug, Default, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct Instant(u64);
```

Comment:
I think it's not proper to put it here since it's not related to `keys`. And I'm afraid we may be confused about many `Instant`s -- the `Instant` here, the `Instant` in `tikv_util` and in `std`.

Reply:
Good point, I'll rename Instant to UnixSecs. I'll move TimeStamp and UnixSecs out into a sub-module. I think they have to be in the keys crate because TimeStamp is used there and both are used in various crates which don't have many deps in common.

Reply:
I prefer to put the `UnixSecs` in `tikv_util/src/time.rs` and put the `TimeStamp` in `components/keys/src/types.rs`